design/src/rfcs/rfc0015_user_friendly_configurable_connectors.md

```diff
@@ -0,0 +1,238 @@
+
+#### The Service Config
+
+For a majority of services, the service config doesn't contain anything more than what was set in the `SdkConfig`, with a notable exception. The service config is responsible for providing a service-specific middleware called the "Default Middleware" that:
+
+- Uses SigV4 to sign requests
+- Attaches credentials to requests
+- Attaces a user agent to requests
+- Attaches info that's used for recursion detection
+- Resolves service endpoints
+
+When calling `aws_sdk_s3::Client::new`, the `SdkConfig` is transformed into an `aws_sdk_s3::Config` by calling an `Into::into` impl.
+
+The service config could also afford users the ability to wrap this middleware or replace it, but that's outside the scope of this RFC.
+
+#### The Service Client
+
+The Service Client wraps the Service Config and the Smithy Client Pool. When used to create a Fluent Builder, it passes on copies of both. In previous versions of smithy-rs, the Service Client was also responsible for creating a Smithy Client, but that responsibility is now held by a Fluent Builder's `send()` method.
+
+### `let resp = client.list_buckets().send().await?;`
+
+Calling `client.list_buckets()` creates a Fluent Builder for the ["list buckets" operation](https://docs.rs/aws-sdk-s3/latest/aws_sdk_s3/operation/struct.ListBuckets.html). The magic happens when `send().await` is called, and that's where knowledge of [RFC-0002](./rfc0002_http_versions.md) is necessary.
+
+1. The Fluent Builder constructs the inner Operation, passing it a copy of the Service Config and the Smithy Client Pool.
+2. All operations have, in the HTTP request Property Bag, a priority list of HTTP versions to use when making requests. That list is fetched from the property bag.
+3. A `MakeConnectorSettings` is created by calling an `Into::into` impl on the Service Config.
+4. Using those settings, and the priority list of HTTP versions, a client is fetched from the Smithy Client Pool.
+5. Finally, the request is sent by calling `client.call(operation).await`
+
+### `let bucket_list = resp.buckets().unwrap();`
+
+There's nothing new to be seen here. The previous line sent the request and recieved the response. This line is just looking at the response.
+
+The inner workings of the Smithy Client Pool
+--------------------------------------------
+
+While the Service Client "owns" the Smithy Client Pool, its only responsibility is to share copies of it with Operations. The client pool has three methods:
+
+- `pub async fn get_or_create_client<E>(&self, make_connector_settings: MakeConnectorSettings, http_versions: &[http::Version]) -> Result<Arc<aws_smithy_client::Client>, SdkError<E>>`
+- `async fn fetch_existing_client(&self, connector_key: &ConnectorKey) -> Option<Arc<aws_smithy_client::Client>>`
+- `async fn initialize_and_store_new_client<E>(&self, connector_key: &ConnectorKey) -> Result<Arc<aws_smithy_client::Client>, SdkError<E>>`
```

Comment:
It seems like these are essentially private functions needed to implement `get_or_create_client`. Is that correct?

Reply:
Yeah, that's correct. You wouldn't call them directly.